iDiff/utils/docker_utils.go

```diff
@@ -0,0 +1,172 @@
+package utils
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"os/exec"
+	"regexp"
+	"strings"
+	"syscall"
+
+	img "github.com/docker/docker/api/types/image"
+	"github.com/docker/docker/client"
+	"github.com/golang/glog"
+)
+
+// ValidDockerVersion determines if there is a Docker client of the necessary version locally installed.
+func ValidDockerVersion() (bool, error) {
+	cli, err := client.NewEnvClient()
+	if err != nil {
+		return false, fmt.Errorf("Docker client error: %s", err)
+	}
+	version := cli.ClientVersion()
+	if version == "1.31" {
+		return true, nil
+	}
+	return false, nil
+}
+
+func GetImageHistory(image string) ([]img.HistoryResponseItem, error) {
+	imageID := image
+	var err error
+	var history []img.HistoryResponseItem
+	if !CheckImageID(image) {
+		imageID, _, err = pullImageCmd(image)
+		if err != nil {
+			return history, err
+		}
+	}
+	histArgs := []string{"history", "--no-trunc", imageID}
+	dockerHistCmd := exec.Command("docker", histArgs...)
+	var response bytes.Buffer
+	dockerHistCmd.Stdout = &response
+	if err := dockerHistCmd.Run(); err != nil {
+		if exiterr, ok := err.(*exec.ExitError); ok {
+			if status, ok := exiterr.Sys().(syscall.WaitStatus); ok && status.ExitStatus() > 0 {
+				glog.Error("Docker History Command Exit Status: ", status.ExitStatus())
+			}
+		} else {
+			return history, err
+		}
+	}
+	history, err = processHistOutput(response)
+	if err != nil {
+		return history, err
+	}
+	return history, nil
+
+}
+
+func processHistOutput(response bytes.Buffer) ([]img.HistoryResponseItem, error) {
+	respReader := bytes.NewReader(response.Bytes())
+	reader := bufio.NewReader(respReader)
+	var history []img.HistoryResponseItem
+	var CreatedByIndex int
+	var SizeIndex int
+	for {
+		var event img.HistoryResponseItem
+		text, _, err := reader.ReadLine()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return history, err
+		}
+
+		line := string(text)
+		if CreatedByIndex == 0 {
+			CreatedByIndex = strings.Index(line, "CREATED BY")
+			SizeIndex = strings.Index(line, "SIZE")
+			continue
+		}
+		event.CreatedBy = line[CreatedByIndex:SizeIndex]
+		history = append(history, event)
+	}
+	return history, nil
+}
+
+func processPullCmdOutput(image string, response bytes.Buffer) (string, string, error) {
+	respReader := bytes.NewReader(response.Bytes())
+	reader := bufio.NewReader(respReader)
+
+	var events []Event
+	for {
+		var event Event
+		text, _, err := reader.ReadLine()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return "", "", err
+		}
+		event.Status = string(text)
+		events = append(events, event)
+	}
+
+	imageDigest, err := getImagePullResponse(image, events)
```

Comment:
Seems like from here down is repetitive code from image_utils? If there's a way to refactor to extract this out into a helper method that can be reused, that'd probably be better...